File: solutions/c/sieve/1/sieve.c
#include "sieve.h"

uint32_t sieve(uint32_t limit, uint32_t *primes, size_t max_primes){
    uint32_t m, n;
    uint32_t count = 0;
    int marks[limit + 1];
    memset(marks, 0, (limit + 1)*sizeof(marks[0]));
    for (uint32_t i = 2; i <= limit; i++){
        if (marks[i]){continue;}
        primes[count++] = i;
        m = i;
        n = 2*m;
        while (n <= limit){
            marks[n] = 1;
            n += m;
        }
    }
    return (count < max_primes) ? count : max_primes;
}

File: solutions/julia/sieve/1/sieve.jl
function sieve(limit)
    numbers = collect(1:limit)
    boolean = fill(true, limit)
    boolean[1], i = false, 1
    while i < limit
        i += 1
        if boolean[i] == false; continue; end
        for multiple in 2i:i:limit
            boolean[multiple] = false
        end
    end
    numbers[boolean]
end